main/nsmiracle/logger.h

```diff
@@ -0,0 +1,164 @@
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+// 3. Neither the name of the University of Padova (SIGNET lab) nor the
+//    names of its contributors may be used to endorse or promote products
+//    derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * @file   logger.h
+ * @author Vincenzo Cimino
+ * @version 1.0.0
+ *
+ * @brief A Logger class represents an object used to print log messages from a
+ * given module of a node.
+ *
+ */
+
+#ifndef LOGGER
+#define LOGGER
+
+#include "object.h"
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+class Logger
+{
+public:
+	/**
+	 * Logger constructor.
+	 *
+	 */
+	Logger();
+
+	/**
+	 * Logger destructor
+	 *
+	 */
+	~Logger();
+
+	/**
+	 * Set the Logger node id.
+	 *
+	 * @param node_id the integer value that will be the logger node id.
+	 *
+	 **/
+	inline void
+	setLogNodeId(size_t node_id)
+	{
+		node_id_ = node_id;
+	}
+
+	/**
+	 * Set the Logger log level.
+	 *
+	 * @param log_level the string representings the amount of logs being
+	 *generated. Valid values are ERR,INFO,DBG.
+	 *
+	 **/
+	inline void
+	setLogLevel(const std::string &log_level)
+	{
+		log_level_ = strToLog(log_level);
+	}
+
+	/**
+	 * Set the Logger file name.
+	 *
+	 * @param log_file the string representings the name of the file where to
+	 *store log messages.
+	 *
+	 **/
+	inline void
+	setLogFile(const std::string &log_file)
+	{
+		log_file_ = log_file;
+	}
+
+	/**
+	 * Method to print the provided log message either to the console
+	 * or to a file if a log_file name is specified.
+	 * The log message is printed only if log_level_ >= log_level.
+	 *
+	 * @param log_level String of valid log level (ERR/INFO/DBG).
+	 * @param module String representing the name of the module.
+	 * @param message String log message.
+	 *
+	 */
+	void printOnLog(const std::string &log_level, const std::string &module,
```

Comment:
why do we need to pass the log_level here, if it can also be set via `setLogLevel`?

Reply:
why we do need to pass it via string and not enum?